[HUDI-1864] Added support for Date, Timestamp, LocalDate and LocalDateTime in TimestampBasedAvroKeyGenerator

vaibhav-sinha opened the pull request:
This PR adds the support for Date, Timestamp, LocalDate and LocalDateTime types in TimestampBasedAvroKeyGenerator. Hence, columns with these types can be used for partitioning. The support for Date and Timestamp types are added because those are used by Spark for DateType and TimestampType respectively. The support for LocalDate and LocalDateTime are added because that is how Avro date, `timestamp-micros` and `timestamp-millis` are represented in `HoodieAvroUtils`. 

In addition to this, `HoodieAvroUtils` did not handle type conversions properly for `timestamp-micros` and `timestamp-millis` logical types and that has been fixed.

Test cases are added to verify the functionality and manual jobs have been run to validate the changes as well.


## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

n3nash commented:
@vaibhav-sinha Looks fine to me. Can you check why the CI is failing ?

nsivabalan commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/keygen/TestTimestampBasedKeyGenerator.java:
```diff
@@ -378,4 +382,46 @@ public void test_ExpectsMatch_MultipleInputFormats_ShortDate_OutputCustomDate()
+  }
+
+  @Test
+  public void test_ExpectsMatch_LogicalType_Timestamp() throws IOException {
```
I see we have added LocalDate, LocalDatetime, Date and Timestamp support to source code, but tests are added only for 2 of these. can we please add for rest as well.

vaibhav-sinha replied:
The tests check for all 4. When GenericRecord is being used to get the partitionPath, then date field goes as LocalDate, whereas when it gets converted to Row, then it is represented as Date. Similarly, for timestamp, when GenericRecord is used, the field is represented as LocalDateTime and when converted to Row, it is represented as Timestamp.

nsivabalan replied:
thanks for clarifying

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/avro/HoodieAvroUtils.java:
```diff
@@ -485,6 +488,10 @@ private static Object convertValueForAvroLogicalTypes(Schema fieldSchema, Object
       }
+    } else if (fieldSchema.getLogicalType() == LogicalTypes.timestampMicros()) {
+      return Instant.EPOCH.plus(Long.parseLong(fieldValue.toString()), ChronoUnit.MICROS).atZone(ZoneId.systemDefault()).toLocalDateTime();
+    } else if (fieldSchema.getLogicalType() == LogicalTypes.timestampMillis()) {
```
can you please point me to testcase which is covering this.

vaibhav-sinha replied:
Have added the test `testGetNestedFieldValForLogicalTypes` in `TestHoodieAvroUtils`

nsivabalan replied:
thanks. looks like we don't have coverage for LogicalTypes.timestampMicros(). have left a comment below.

vaibhav-sinha replied:
Added

danny0405 replied:
Why we also change the value of type `LogicalTypes.timestampMillis` from long to `LocalDataTime` ? What is the gain ?

vaibhav-sinha replied:
We have just tried to maintain consistent data types for logical types. Previously, LogicalTypes.date was represented as LocalDate but LogicalTypes.timestampMillis was represented at long.

nsivabalan commented on hudi-common/src/test/java/org/apache/hudi/avro/TestHoodieAvroUtils.java:
```diff
@@ -79,6 +82,18 @@
+      + "    {\"name\": \"field1\", \"type\": [\"null\", \"string\"], \"default\": null},\n"
+      + "    {\"name\": \"createTime\", \"type\": [\"null\", \"long\"], \"default\": null},\n"
+      + "    {\"name\": \"dob\", \"type\": {\"type\": \"int\" ,\"logicalType\": \"date\"}},\n"
+      + "    {\"name\": \"updatedAt\", \"type\": {\"type\": \"long\", \"logicalType\": \"timestamp-millis\"}}\n"
```
Can you add a field for timestamp-micros as well and enhance testGetNestedFieldValForLogicalTypes for the same.

vaibhav-sinha replied:
Added

nsivabalan reviewed (commented):
minor comment. LGTM. Once you address it, I can merge.

vaibhav-sinha commented:
Flink testcase was failing because conversion to Flink TimestampData was not handled. Have pushed a fix for that as well.

nsivabalan commented:
Can you check CI failure please.

vaibhav-sinha commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/keygen/KeyGenUtils.java:
```diff
@@ -62,7 +62,7 @@
           return "";
         } else {
-          return kvArray[1];
+          return String.join(":", Arrays.copyOfRange(kvArray, 1, kvArray.length));
```
In case when one of keys in a complex key has a value which contain `:` character, the earlier implementation would not extract the correct value since it assumes the format is `key:value` with value not containing any `:` character itself. This is a problem in general, and also specifically if timestamp is used for partitioning because it will have `:` character in the time part.

n3nash replied:
Seems like a correct fix.

vaibhav-sinha commented:
The tests were clean except for one test case failing before which I had fixed. But after merging the latest changes from master, I see a lot of tests failing and the errors seem unrelated to this change. I have again merged the latest changes. Can you retrigger the build.

nsivabalan commented on hudi-flink/src/test/java/org/apache/hudi/sink/TestWriteCopyOnWrite.java:
```diff
@@ -380,12 +380,12 @@ public void testUpsertWithDelete() throws Exception {
   public void testInsertWithMiniBatches() throws Exception {
     // reset the config option
-    conf.setDouble(FlinkOptions.WRITE_BATCH_SIZE, 0.0006); // 630 bytes batch size
+    conf.setDouble(FlinkOptions.WRITE_BATCH_SIZE, 0.00075); // 786 bytes batch size
```
are these the tests you were mentioning were failing in CI? ideally these tests should not be impacted by this patch right?

vaibhav-sinha replied:
Yes. I ran the tests from the master branch on my local and they seem to fail for these as well. What I could figure figure out was the size estimate of the HoodieRecord calculated by HoodieRecordSizeEstimator was higher from what was being assumed in the testcase. Hence the batch size was being set to a lower value in the testcase causing the flush to happen earlier than expected. I have modified the batch size based on the output of HoodieRecordSizeEstimator. Not sure if this is the right fix though.

nsivabalan replied:
hmmm, strange. I see quite a few PRs shows green. so unlikely this fails consistently. also, if there are any such unrelated failures, lets isolate them to separate PR. I will try to reproduce the issue locally in master as well.

n3nash replied:
This does seem unrelated. @nsivabalan Please try to reproduce this before landing this.

nsivabalan replied:
@danny0405 @leesf @yanghua : hey folks. Can one of you check why these tests are failing for this patch and why the fix is required. I also tried locally and it is failing w/ this patch, but don't have lot of context around tests. I expected these tests should not be affected by the changes in this patch.

nsivabalan replied:
Also, nishith left a comment else where about backwards compatability. Can you check that out as well.

danny0405 replied:
Thanks for the reminder, the `TestWriteCopyOnWrite#testInsertWithMiniBatches` actually test flushing data buffer based on the buffer size, the `StreamWriteFunction` detect the record buffer size using Java Reflection, does this patch affect the record size (for e.g, the data structure of the record) ?

danny0405 replied:
Is it related with the JDK version ? What JDK version did you use @vaibhav-sinha ?

vaibhav-sinha replied:
I used jdk1.8.0_171.jdk.

vaibhav-sinha replied:
I think the reason for this is as follows:

For timestamp field, HoodieAvroUtils now returns an instance of LocalDateTime instead of Long. When Row data is being converted to HoodieRecordPayload, then `orderingVal` is being fetched using `HoodieAvroUtils.getNestedFieldVal`. Hence the structure of the payload has changed and is of larger size than before. 

```
public HoodieRecordPayload<?> createPayload(GenericRecord record, boolean isDelete) throws Exception {
    if (shouldCombine) {
      ValidationUtils.checkState(preCombineField != null);
      Comparable<?> orderingVal = (Comparable<?>) HoodieAvroUtils.getNestedFieldVal(record,
          preCombineField, false);
      return (HoodieRecordPayload<?>) constructor.newInstance(
          isDelete ? null : record, orderingVal);
    } else {
      return (HoodieRecordPayload<?>) this.constructor.newInstance(Option.of(record));
    }
  }
```

danny0405 replied:
Okey, got your idea, can we keep the type as long if the data type is timestamp without time zone (or UTC timezone), we can keep the compatibility and the long has less memory footprint as you have tested.

This is important because Hoodie writer handle buffer the records, so the less memory the better.

nsivabalan commented:
@vaibhav-sinha : Not sure how you rebased with master, but I see your commits are interleaved with commits in master. Also, I tried running TestWriteCopyOnWrite tests in apache master and its succeeding. 
Would you mind putting up a new PR with just the required changes from this PR. Start with a new branch from latest master. And we can go from there. 

And whenever you wanna rebase your feature branch w/ master, do the following.
1. ensure you are in your feature branch. I assume this is your own repo with upstream as apache/hudi
2. git fetch upstream master
3. git rebase upstream/master

So, whenever you fetch and rebase, your commits will be rebased on top of latest master. in other words, all commits in your feature branch should be on top of latest master commit once. your commits should not be interleaved.

n3nash commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/keygen/TestTimestampBasedKeyGenerator.java:
```diff
@@ -378,4 +382,46 @@ public void test_ExpectsMatch_MultipleInputFormats_ShortDate_OutputCustomDate()
   }
+
+  @Test
+  public void test_ExpectsMatch_LogicalType_Date() throws IOException {
```
Can we use camelcase please ? Let's follow the coding patterns used in the project.

vaibhav-sinha replied:
Had named the test based on the naming convention for existing tests in this class. Have updated names of all tests to camelCase.

n3nash commented on hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/keygen/TestTimestampBasedKeyGenerator.java:
```diff
@@ -378,4 +382,46 @@ public void test_ExpectsMatch_MultipleInputFormats_ShortDate_OutputCustomDate()
+  }
+
+  @Test
+  public void test_ExpectsMatch_LogicalType_Timestamp() throws IOException {
```
Same here

vaibhav-sinha replied:
Done

n3nash commented on hudi-flink/src/main/java/org/apache/hudi/util/StringToRowDataConverter.java:
```diff
@@ -85,7 +87,8 @@ private static Converter getConverter(LogicalType logicalType) {
       case TIMESTAMP_WITHOUT_TIME_ZONE:
-        return field -> TimestampData.fromEpochMillis(Long.parseLong(field));
+        // see HoodieAvroUtils#convertValueForAvroLogicalTypes
+        return field -> TimestampData.fromEpochMillis(LocalDateTime.parse(field).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli());
```
Is this change backwards compatible ?

danny0405 replied:
Why we need this change, is it related with this patch ? And the change is wrong because the string format is already in epoch milliseconds but not the format `DateTimeFormatter.ISO_LOCAL_DATE_TIME`.

vaibhav-sinha replied:
Yes it is related to this patch. In HoodieAvroUtils, whenever the LogicalType is timestamp, LocalDateTime instance is being returned instead of Long. Hence in string form, we would get a DateTime string in StringToRowConverter instead of stringified Long as before. Hence it needs to be parsed using LocalDateTime parser. Check the handling of LocalDate in the line above which was there from before because LogicalType date was handled correctly in HoodieAvroUtils before only.

n3nash reviewed (commented):
@nsivabalan @vaibhav-sinha Few comments in addition to what's remaining.

nsivabalan commented:
@vaibhav-sinha : in the mean time, can you address nishith's feedback on test naming.

nsivabalan commented:
@vaibhav-sinha : I recreated this PR cleanly [here](https://github.com/apache/hudi/pull/3039). Have not made fixes to TestCopyOnWrite yet. will wait for flink experts to chime in.

danny0405 commented on hudi-flink/src/main/java/org/apache/hudi/sink/utils/PayloadCreation.java:
```diff
@@ -76,6 +78,9 @@ public static PayloadCreation instance(Configuration conf) throws Exception {
       Comparable<?> orderingVal = (Comparable<?>) HoodieAvroUtils.getNestedFieldVal(record,
           preCombineField, false);
+      if(orderingVal instanceof LocalDateTime) {
+        orderingVal = ((LocalDateTime) orderingVal).atZone(ZoneId.systemDefault()).toInstant().toEpochMilli();
```
Why not just use a long if the `orderingVal` data type has no timezone.

vaibhav-sinha replied:
The data types returned from `HoodieAvroUtils.getNestedFieldVal` can be any valid Avro types including string, int, decimal, uuid, etc. How do you to check for timezone in these?

danny0405 replied:
Then what is the theory to use `ZoneId.systemDefault()` ?

vaibhav-sinha replied:
When converting epoch to `LocalDateTime`, a timezone needs to be specified, and we are using `ZoneId.systemDefault()` there. Hence when converting back from `LocalDateTime` to epoch, we are using the same.

danny0405 commented:
This change increase the memory footprint of `HoodieRecord` if the ordering key is timestamp(3) type without time zone, which is a regression for hoodie writer, can we avoid that ?

vaibhav-sinha commented:
If we don't want to impact the memory footprint of `HoodieRecord`,  we have two options:

1. We could convert the value from `timestamp` to `long` when setting `orderingVal` at all places.
2. We do not represent `timestamp` based `LogicalTypes` as `LocalDateTime`.

@nsivabalan Any suggestions on which approach should we go ahead with?

nsivabalan commented:
@n3nash : Can you please respond to the author.

n3nash commented:
@vaibhav-sinha For your second suggestion, what would we represent `timestamp` based LogicalType as instead of `LocalDateTime` ? Are you suggesting using long ?

vaibhav-sinha commented:
@n3nash Essentially we did two things in this PR. We changed the representation of `timestamp` based LogicalType and added the support for Data/Timestamp based objects in `TmestampBasedAvroKeyGenerator`. Both of these are independent of each other.

Hence, we can continue using `long` to represent `timestamp` if the memory footprint is a concern. In that case, I would suggest we should represent `Date` as `integer` just to be consistent with representations.

codope reviewed (changes requested):
@vaibhav-sinha The changes are in good shape now. Could you please resolve the merge conflicts?

xushiyan commented:
https://github.com/apache/hudi/pull/3039
